### PR TITLE
[TIR] Add cp.async support for tir.if_then_else

### PR DESCRIPTION
**Pull request by Rainy-Memory**
**Description**

This PR supports CUDA's cp.async ptx for un-vectorized BufferStore from a `tir.if_then_else` call and thus enables padded async copy.

**Modifications**

1. Add a test script that shows the performance boost brought by this PR.
2. Update `PTXAsyncCopyInjector` Pass to support async copy on an `if_then_else` call.
3. Update other support files to print the corresponding ptx.

**Performance**

The unittest this PR included shows an example using async copy on a padded copy. The performance that enables or not this PR is shown below: (in the unit of seconds)

|Mainline TVM|Enable this PR|
|-|-|
|3.4076e-6|2.4371e-6|

The integration test shows a GEMM workload M = 1012, N = 1017, K=1014 padded to MNK=1024 and using async copy. CUDA Core performance is shown below: (in the unit of GFLOP/s)

|Mainline TVM|Enable this PR|
|-|-|
|5152.18|14780.49|

(Running in RTX 3080)

### CODE REVIEW

File: src/target/source/codegen_cuda.cc [hunk 1]
```diff
@@ -914,7 +914,12 @@ void CodeGenCUDA::VisitExpr_(const CallNode* op, std::ostream& os) {
     std::string src = this->PrintExpr(op->args[2]);
     std::string src_offset = this->PrintExpr(op->args[3]);
     std::string size = this->PrintExpr(op->args[4]);
-    this->stream << PrintCpAsyncAssembly(dst, dst_offset, src, src_offset, size);
+    // use size of argument list to indicate whether or not to use predicated cp.async
+    if (op->args.size() == 5)
+      this->stream << PrintCpAsyncAssembly(dst, dst_offset, src, src_offset, size);
+    else
+      this->stream << PrintPredicatedCpAsyncAssembly(dst, dst_offset, src, src_offset,
+                                                     size, this->PrintExpr(op->args[5]));
```
Comment (junrushao) [inline src/target/source/codegen_cuda.cc:922]: nit: always use `{}`

```suggestion
    if (op->args.size() == 5) {
      this->stream << PrintCpAsyncAssembly(dst, dst_offset, src, src_offset, size);
    }
    else {
      this->stream << PrintPredicatedCpAsyncAssembly(dst, dst_offset, src, src_offset,
                                                     size, this->PrintExpr(op->args[5]));
    }
```

File: src/tir/transforms/inject_ptx_async_copy.cc [hunk 2]
```diff
@@ -47,74 +47,104 @@ class PTXAsyncCopyInjector : public StmtMutator {
...
-              ICHECK(store->buffer.scope() == "shared.dyn");
-              ICHECK(dst_elem_type.value() == DataType::UInt(8));
-              // BufferStore/Load have the "pointer reinterpret" semantics according to their
-              // "value" dtype. Their "indices" are supposed to be applied after such pointer cast,
-              // for example: ((*float16)(byte_buffer))[buffer->indices] = fp16_value;
-              // To replace BufferStore/Load with cp.async, we need to multiply the store index by
-              // the byte size of the "value" dtype, to get the correct offset into the byte buffer.
-              index_factor = src_elem_type->bytes();
-            }
+  Stmt injectPTX(const BufferLoadNode* load, const BufferStoreNode* store,
```
Comment (junrushao) [inline src/tir/transforms/inject_ptx_async_copy.cc:50]: nit: style

```suggestion
  Stmt InjectPTX(const BufferLoadNode* load, const BufferStoreNode* store,
```

File: tests/python/unittest/test_cp_async_in_if_then_else.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+                                1024,
+                                ax0_0_ax1_0_fused % 8 * 128
+                                + ax0_1_ax1_1_fused % 2 * 64
+                                + ax1_2_0_ax0_2_0_ax0_2_1_ax1_2_1_ax0_2_2_fused % 32 // 2 * 4
+                                + ax1,
+                            )
+                            T.reads(Y_reindex_local[v0, v1])
+                            T.writes(Y[v0, v1])
+                            Y[v0, v1] = Y_reindex_local[v0, v1]
+
```
Comment (junrushao) [inline tests/python/unittest/test_cp_async_in_if_then_else.py:284]: can we assert the cuda code generated btw?

File: tests/python/unittest/test_cp_async_in_if_then_else.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+def test_matmul():
+    with tvm.transform.PassContext(config={"tir.use_async_copy": 1}):
+        rt_mod = tvm.build(Module, target="cuda")
```
Comment (junrushao) [inline tests/python/unittest/test_cp_async_in_if_then_else.py:288]: are async copy instructions always available? particularly on previous generations of cuda gpus

Comment (Rainy-Memory) [reply tests/python/unittest/test_cp_async_in_if_then_else.py:294]: now compute version check is added to the unittest

**Comment by junrushao** [comment]
the shape is:

```
A: T.Buffer[(1012, 1014), "float32"],
B: T.Buffer[(1014, 1017), "float32"],
C: T.Buffer[(1012, 1017), "float32"],
```

let's add ^ this to the PR description.

**Comment by junrushao** [comment]
Please fix the lint and tests so that we could get it in ASAP

**Review by junrushao — APPROVED** [review]
LGTM! CC @tqchen for a second look

**Comment by junrushao** [comment]
@Rainy-Memory Hey sorry for the late followup, but I just realized...would you like to merge the unittest `test_cp_async_in_if_then_else` into `tests/python/unittest/test_tir_transform_inject_ptx_async_copy.py`?

**Pull request merged**